### PR TITLE
Tdr 3264 move triggering backend checks

### PR DESCRIPTION
**Pull request by thanhz**
This removes updatingconsignmentstatus and backendchecks entirely from the upload.js to the application.

### CODE REVIEW

File: app/controllers/FileChecksController.scala [hunk 1]
```diff
@@ -37,39 +43,70 @@ class FileChecksController @Inject() (
...
+      }
+    } yield result
+  }
+
+  private def handleSuccessfulUpload(consignmentId: UUID, reference: String, isJudgmentUser: Boolean)(implicit request: Request[AnyContent]) = {
+    val token = request.token.bearerAccessToken
+    (for {
+      statuses <- consignmentStatusService.getConsignmentStatuses(consignmentId, token)
+      uploadStatus = statuses.find(_.statusType == UploadType.id)
+      alreadyTriggered = uploadStatus.isDefined && uploadStatus.get.value == CompletedValue.value || uploadStatus.get.value == CompletedWithIssuesValue.value
```
Comment (TomJKing) [inline app/controllers/FileChecksController.scala:102]: If the status is `CompletedWithIssues` does the user still see an error message on the page?

File: app/controllers/FileChecksController.scala [hunk 2]
```diff
@@ -78,12 +115,19 @@ class FileChecksController @Inject() (
...
       } else {
-        Ok(views.html.judgment.judgmentFileChecksProgress(consignmentId, reference, applicationConfig.frontEndInfo, request.token.name)).uncache()
+        if (isJudgmentUser) {
+          Ok(views.html.judgment.judgmentFileChecksProgress(consignmentId, reference, applicationConfig.frontEndInfo, request.token.name)).uncache()
+        } else {
+          Ok(views.html.standard.fileChecksProgress(consignmentId, reference, applicationConfig.frontEndInfo, request.token.name))
+            .uncache()
+        }
       }
+    }).recover { case exception: Exception =>
```
Comment (TomJKing) [inline app/controllers/FileChecksController.scala:129]: Does the user still see an error on the page if this is the case?

File: app/controllers/FileChecksController.scala [hunk 3]
```diff
@@ -37,39 +43,70 @@ class FileChecksController @Inject() (
...
-            request.token.name,
-            isJudgmentUser = false
-          )
-        ).uncache()
-      } else {
-        Ok(views.html.standard.fileChecksProgress(consignmentId, reference, applicationConfig.frontEndInfo, request.token.name))
-          .uncache()
+      reference <- consignmentService.getConsignmentRef(consignmentId, token)
+      result <- uploadFailed match {
+        case Some(TRUE) => handleFailedUpload(consignmentId, reference, isJudgmentUser = false, token)
```
Comment (TomJKing) [inline app/controllers/FileChecksController.scala:80]: Don't use the apache `TRUE`. Just use `case Some("true") ...` Or write a common utils method to convert a string option to boolean.

File: app/controllers/FileChecksController.scala [hunk 4]
```diff
@@ -37,39 +43,70 @@ class FileChecksController @Inject() (
...
 
-  def judgmentFileChecksPage(consignmentId: UUID): Action[AnyContent] = judgmentTypeAction(consignmentId) { implicit request: Request[AnyContent] =>
+  def judgmentFileChecksPage(consignmentId: UUID, uploadFailed: Option[String]): Action[AnyContent] = judgmentTypeAction(consignmentId) { implicit request: Request[AnyContent] =>
+    val token = request.token.bearerAccessToken
     for {
-      fileChecks <- getFileChecksProgress(request, consignmentId)
-      reference <- consignmentService.getConsignmentRef(consignmentId, request.token.bearerAccessToken)
+      reference <- consignmentService.getConsignmentRef(consignmentId, token)
+      result <- uploadFailed match {
+        case Some(TRUE) => handleFailedUpload(consignmentId, reference, isJudgmentUser = true, token)
```
Comment (TomJKing) [inline app/controllers/FileChecksController.scala:91]: Don't use the apache `TRUE`. Just use `case Some("true")` ... Or write a common utils method to convert a string option to boolean.

File: app/controllers/UploadController.scala [hunk 5]
```diff
@@ -90,7 +74,7 @@ class UploadController @Inject() (
       transferAgreementStatus match {
         case Some(CompletedValue.value) =>
           uploadStatus match {
-            case Some(InProgressValue.value) =>
+            case Some(InProgressValue.value) | Some(CompletedWithIssuesValue.value) =>
```
Comment (TomJKing) [inline app/controllers/UploadController.scala:77]: Will this still show an error on the page if the status is `CompletedWithIssuesValue`?

File: app/controllers/UploadController.scala [hunk 6]
```diff
@@ -124,7 +108,7 @@ class UploadController @Inject() (
       val pageHeadingUploading = "Uploading judgment"
 
       uploadStatus match {
-        case Some(InProgressValue.value) =>
+        case Some(InProgressValue.value) | Some(CompletedWithIssuesValue.value) =>
```
Comment (TomJKing) [inline app/controllers/UploadController.scala:111]: Will this still show an error on the page if the status is `CompletedWithIssuesValue`?

File: npm/test/upload.test.ts [hunk 7]
```diff
@@ -104,30 +70,7 @@ const mockUploadFailure: () => void = () => {
...
-  expect(consoleErrorSpy).not.toHaveBeenCalled()
-  expect(mockGoToNextPage).toHaveBeenCalled()
-
-  mockGoToNextPage.mockRestore()
-  consoleErrorSpy.mockRestore()
-})
-
-test("upload function throws an error when upload fails", async () => {
-  fetchMock.mockResponse(JSON.stringify({ updateConsignmentStatus: 1 }))
+test("upload function will redirect to the next page with uploadFailed set to true if the upload fails", async () => {
```
Comment (TomJKing) [inline npm/test/upload.test.ts:73]: Could you reword the test, to make it specific which page is redirected to.

File: npm/test/upload.test.ts [hunk 8]
```diff
@@ -138,15 +81,13 @@ test("upload function throws an error when upload fails", async () => {
...
 
-  expect(mockGoToNextPage).not.toHaveBeenCalled()
+  expect(mockGoToNextPage).lastCalledWith("12345", "true", false)
 
   mockGoToNextPage.mockRestore()
 })
 
-test("upload function will not redirect to the next page if the backend checks trigger fails", async () => {
-  fetchMock.mockResponse(JSON.stringify({ updateConsignmentStatus: 1 }))
+test("upload function redirects to the next page with uploadFailed set to false if the upload succeeds", async () => {
```
Comment (TomJKing) [inline npm/test/upload.test.ts:89]: Could you reword the test, to make it specific which page is redirected to.

File: test/controllers/FileChecksControllerSpec.scala [hunk 9]
```diff
@@ -81,33 +89,42 @@ class FileChecksControllerSpec extends FrontEndTestHelper with TableDrivenProper
...
           graphQLConfiguration,
           keycloakConfiguration,
           consignmentService,
-          frontEndInfoConfiguration
+          frontEndInfoConfiguration,
+          backendChecksService,
+          consignmentStatusService
         )
-
+        val uploadFailed = FALSE
```
Comment (TomJKing) [inline test/controllers/FileChecksControllerSpec.scala:119]: Don't use the apache library `FALSE`, use `Some(False)`

File: test/controllers/FileChecksControllerSpec.scala [hunk 10]
```diff
@@ -159,29 +184,45 @@ class FileChecksControllerSpec extends FrontEndTestHelper with TableDrivenProper
...
           consignmentService,
-          frontEndInfoConfiguration
+          frontEndInfoConfiguration,
+          backendChecksService,
+          consignmentStatusService
         )
-
-        val fileChecksCompletePage = if (userType == "judgment") {
-          controller.judgmentFileChecksPage(consignmentId).apply(FakeRequest(GET, s"/$pathName/$consignmentId/file-checks"))
+        val uploadFailed = FALSE
```
Comment (TomJKing) [inline test/controllers/FileChecksControllerSpec.scala:210]: Don't use the apache library `FALSE`, use `Some(False)`

File: test/controllers/FileChecksControllerSpec.scala [hunk 11]
```diff
@@ -1,30 +1,39 @@
...
-import com.github.tomakehurst.wiremock.client.WireMock.{status, _}
+import com.github.tomakehurst.wiremock.client.WireMock._
 import configuration.GraphQLConfiguration
+import graphql.codegen.GetConsignmentStatus.getConsignmentStatus.GetConsignment.ConsignmentStatuses
 import graphql.codegen.GetFileCheckProgress.{getFileCheckProgress => fileCheck}
+import graphql.codegen.UpdateConsignmentStatus.updateConsignmentStatus
 import io.circe.Printer
 import io.circe.generic.auto._
 import io.circe.syntax._
+import org.apache.commons.lang3.BooleanUtils.{FALSE, TRUE}
```
Comment (TomJKing) [inline test/controllers/FileChecksControllerSpec.scala:12]: Don't use this apache library.

File: test/controllers/UploadControllerSpec.scala [hunk 12]
```diff
@@ -312,12 +310,88 @@ class UploadControllerSpec extends FrontEndTestHelper {
...
       checkForExpectedPageContentOnOtherUploadPages(uploadPageAsString, uploadStatus = uploadStatus)
       uploadPageAsString must include(
-        s"""      <a href="/consignment/$consignmentId/file-checks" role="button" draggable="false" class="govuk-button govuk-button--primary">
+        s"""      <a href="/consignment/$consignmentId/file-checks?uploadFailed=false" role="button" draggable="false" class="govuk-button govuk-button--primary">
            |        Continue
            |      </a>""".stripMargin
       )
     }
 
+    "render the 'upload in progress' page if a standard upload is in completedWithIssues" in {
```
Comment (TomJKing) [inline test/controllers/UploadControllerSpec.scala:319]: Maybe reword test slightly to: `... the 'upload in progress' error page ... has a 'completedWithIssues' status`

File: test/controllers/UploadControllerSpec.scala [hunk 13]
```diff
@@ -312,12 +310,88 @@ class UploadControllerSpec extends FrontEndTestHelper {
...
+        .apply(FakeRequest(GET, s"/consignment/$consignmentId/upload").withCSRFToken)
+      val uploadPageAsString = contentAsString(uploadPage)
+
+      status(uploadPage) mustBe OK
+
+      checkPageForStaticElements.checkContentOfPagesThatUseMainScala(uploadPageAsString, userType = "standard")
+      checkForExpectedPageContentOnOtherUploadPages(uploadPageAsString, uploadStatus = CompletedWithIssuesValue.value)
+    }
+
+    "render the 'upload in progress error' page if a judgment file upload is in CompletedWithIssues" in {
```
Comment (TomJKing) [inline test/controllers/UploadControllerSpec.scala:357]: Maybe reword test name to: `... has a 'completedWithIssues' status`

**Review by TomJKing — COMMENTED** [review]
Looks good. Some initial comments and questions. As its a pretty big PR might need a couple of reviews.

File: app/controllers/UploadController.scala (hunk 5 above)
Comment (thanhz) [reply app/controllers/UploadController.scala:77]: Yes it will. This is extra case is for when the user clicks "back" on the their browser but the backend checks failed they won't see the "something went wrong page"

File: app/controllers/UploadController.scala (hunk 6 above)
Comment (thanhz) [reply app/controllers/UploadController.scala:111]: Yes it will

File: app/controllers/FileChecksController.scala (hunk 2 above)
Comment (thanhz) [reply app/controllers/FileChecksController.scala:128]: Yes, any exception that is thrown will be caught and user is taken to the error page.

File: app/controllers/FileChecksController.scala (hunk 1 above)
Comment (thanhz) [reply app/controllers/FileChecksController.scala:101]: Ah no it doesn't. I've made some changes to catch this.

File: npm/src/index.ts [hunk 14]
```diff
@@ -121,9 +110,12 @@ export const renderModules = async () => {
     } else {
       errorHandlingModule.handleUploadError(frontEndInfo)
     }
-  } else if (timeoutDialog) {
-    const sessionTimeoutModule = await import("./auth/session-timeout")
-    await sessionTimeoutModule.initialiseSessionTimeout()
+  }
+  if (!uploadContainer && !fileChecksContainer) {
```
Comment (TomJKing) [inline npm/src/index.ts:114]: The `timeoutDialog`is no longer appearing on the upload page. I think it has something to do with this change. Was there a reason you added the `if(!uploadContainer && ! fileChecksContainer)`?

**Review by TomJKing — COMMENTED** [review]
Additional comment following some local testing.

File: npm/src/index.ts (hunk 14 above)
Comment (thanhz) [reply npm/src/index.ts:114]: Yes, when on the uploading page if the dialog appears and the 5 minutes has passed or upload succeeds. The prompt Are you sure you want to navigate _away from this page_ appears preventing the page from redirecting unless a user interacts. I tried to disable this no avail.

Comment (TomJKing) [reply npm/src/index.ts:114]: This is the current behaviour on intg. So we should add the session dialog box to the pages. The important thing is that a user is logged out if they do not confirm they want to remain logged in.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
